cli/azd/pkg/input/previewer.go

```diff
@@ -0,0 +1,89 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package input
```

Comment:
All external tools (like terraform, bicep, docker, etc.) depends on the `CommandRunner`.
Each tool has an implementation for how to use the commandRunner to execute.

My proposal/design here is that if a tool wants to show its output with a Previewer, the tool only needs to require the `input.Console` to be created. By making the tool's constructor to take not only a `commandRunner` but also `input.Console`, the tool is making clear about the intentions of defining not only how to invoke the tool, but also what to do with the output.
Having the `input.Console`, the tool uses the CommandRunnerArgs to set the PreviewerOptions, in case the tool is expecting to show a previewer.

We don't have a requirement for letting customers to decide when and/or how to display the external tool's output.
We are currently using each tool's design+implementation to define how to call/invoke them.
For example, when using gitCli, its implementation knows and defines what commands must be called as interactive. Like:

![image](https://github.com/Azure/azure-dev/assets/24213737/6fad8155-8742-4805-be52-e13901361d01)

Following the same approach, I am expecting that each tool will define what commands are expected to be ran using a previewer. Since the tool is the only one who knows about each command and the type of output supported.

In case of terraform, the previewer can't be used, for example, because terraformCLI might require user interaction to confirm or provide values. That's why we define `WithInteractive` mode:

![image](https://github.com/Azure/azure-dev/assets/24213737/ae365d06-e614-434d-ab2a-2832d3b74f4c)

In the future, if we ever want to make the Previewer to support interacting with the user, I would expect the Console to be the orchestrator there.

Reply:
I definitely prefer for us to optimize for displaying output for external tools, and less about a messaging system where messages can be sent arbitrarily anywhere.